simulator/core/perf_sim.cpp

```diff
@@ -180,22 +231,45 @@ void PerfSim<ISA>::clock_decode( Cycle cycle)
 
     auto instr = read_instr( cycle);
 
-    /* TODO: replace all this code by introducing Forwarding unit */
-    if( rf->check_sources( instr))
-    {
-        rf->read_sources( &instr);
+    if ( bypassing_unit->is_stall( instr))
+    {   
+        // data hazard, stalling pipeline
+        wp_decode_2_fetch_stall->write( true, cycle);
+        wp_decode_2_decode->write( instr, cycle);
+        sout << instr << " (data hazard)\n";
+        return;   
+    }
 
-        wp_decode_2_execute->write( instr, cycle);
+    const auto src1_reg_num = instr.get_src1_num();
+    const auto src2_reg_num = instr.get_src2_num();
 
-        /* log */
-        sout << instr << std::endl;
+    if ( bypassing_unit->is_in_RF( src1_reg_num))
+    {
+        rf->read_source_1( &instr);
```

Comment:
You may remove duplicated code by adding `index` field to DataBypass and RF methods
```c++
rf->read_source( &instr, 0);
rf->read_source( &instr, 1);
```